ddtrace/contrib/trace_utils_llm.py

```diff
@@ -0,0 +1,135 @@
+import abc
+import os
+import time
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import TYPE_CHECKING
+
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.contrib.trace_utils import int_service
+from ddtrace.internal.dogstatsd import get_dogstatsd_client
+from ddtrace.internal.hostname import get_hostname
+from ddtrace.internal.log_writer import V2LogWriter
+from ddtrace.sampler import RateSampler
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    from ddtrace import Pin
+    from ddtrace import Span
+
+
+class BaseLLMIntegration:
+    _integration_name = "baseLLM"
+
+    def __init__(self, config, stats_url, site, api_key):
+        self._statsd = get_dogstatsd_client(stats_url, namespace=self._integration_name)
+        self._config = config
+        self._log_writer = V2LogWriter(
+            site=site,
+            api_key=api_key,
+            interval=float(os.getenv("_DD_%s_LOG_WRITER_INTERVAL" % self._integration_name.upper(), "1.0")),
+            timeout=float(os.getenv("_DD_%s_LOG_WRITER_TIMEOUT" % self._integration_name.upper(), "2.0")),
+        )
+        self._span_pc_sampler = RateSampler(sample_rate=config.span_prompt_completion_sample_rate)
+        self._log_pc_sampler = RateSampler(sample_rate=config.log_prompt_completion_sample_rate)
+
+    def is_pc_sampled_span(self, span):
+        # type: (Span) -> bool
+        if not span.sampled:
+            return False
+        return self._span_pc_sampler.sample(span)
+
+    def is_pc_sampled_log(self, span):
+        # type: (Span) -> bool
+        if not self._config.logs_enabled or not span.sampled:
+            return False
+        return self._log_pc_sampler.sample(span)
+
+    def start_log_writer(self):
+        # type: (...) -> None
+        self._log_writer.start()
+
+    @abc.abstractmethod
+    def _set_base_span_tags(self, span):
+        # type: (Span) -> None
+        """Set default LLM span attributes when possible."""
+        pass
+
+    def trace(self, pin, operation_id):
+        # type: (Pin, str) -> Span
+        """
+        Start a LLM request span.
+        Reuse the service of the application since we'll tag downstream request spans with the LLM name.
+        Eventually those should also be internal service spans once peer.service is implemented.
+        """
+        span = pin.tracer.trace(
+            "%s.request" % self._integration_name, resource=operation_id, service=int_service(pin, self._config)
```

Comment:
Can llm span be external? Ex: Wraps an outbound request or a database call

Reply:
Nope I don't believe so.